src/prefect/agent/kubernetes/agent.py

```diff
@@ -93,9 +95,139 @@ def __init__(
+        _continue = ""
+        while more:
+            try:
+                jobs = self.batch_client.list_namespaced_job(
+                    namespace=self.namespace,
+                    label_selector="prefect.io/identifier",
+                    limit=20,
+                    _continue=_continue,
+                )
+                _continue = jobs.metadata._continue
+                more = bool(_continue)
+
+                for job in jobs.items:
+                    pods = self.core_client.list_namespaced_pod(
+                        namespace=self.namespace,
+                        label_selector="prefect.io/identifier={}".format(
+                            job.metadata.labels.get("prefect.io/identifier")
+                        ),
+                    )
+
+                    pod_names = []
+                    for pod in pods.items:
+                        pod_names.append(pod.metadata.name)
+
+                    jobs_list.append(
+                        {
+                            "job_name": job.metadata.name,
+                            "pod_names": pod_names,
+                            "flow_run_id": job.metadata.labels.get(
+                                "prefect.io/flow_run_id"
+                            ),
+                        }
+                    )
+            except self.k8s_client.rest.ApiException as exc:
+                if exc.status == 410:
+                    self.logger.debug("List jobs continue token expired, relisting")
+                    _continue = ""
+                    continue
+
+        self.logger.debug(
+            f"Found {len(jobs_list)} existing prefect jobs in the cluster"
+        )
+        return jobs_list
+
+    def heartbeat(self) -> None:
+        """
+        Check status of jobs created by this agent. This function checks if jobs are `Failed` or
+        `Succeeded` and if they are then the jobs are deleted from the namespace. If one of the job's
+        pods happen to run into image pulling errors then the flow run is failed and the job is still
+        deleted.
+        """
+
+        self.logger.debug(f"Reading statuses for {len(self.jobs)} jobs...")
+        for job in copy.deepcopy(self.jobs):
+            delete_job = False
+
+            try:
+                job_status = self.batch_client.read_namespaced_job_status(
+                    namespace=self.namespace, name=job["job_name"],
+                ).status
+            except self.k8s_client.rest.ApiException:
+                delete_job = True
+
+            if not delete_job and not job_status.failed and not job_status.succeeded:
+                for pod_name in job["pod_names"]:
+                    try:
+                        pod_status = self.core_client.read_namespaced_pod_status(
+                            namespace=self.namespace, name=pod_name,
+                        ).status
+                    except self.k8s_client.rest.ApiException:
+                        delete_job = True
```

Comment:
There's a few reasons we might get an api exception here (unknown pod id, auth issue, network error, ...), I think we'll want to be more specific when determining if that failure should cause the backing job to be deleted.